### ISSUE TITLE
support arbitrary-precision types

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#290]
**Reported by rcurtin on 6 Nov 43643786 13:21 UTC**
Right now everything works on double and arma::mat.  Based on #300, it would be useful to support arbitrary-precision types (within the confines of what Armadillo offers).  This ticket right now just proposes the ideas.  There will be numerous difficult implementation details that may make this very difficult in the end; those can be detailed in the comments as they come up.

Migrated-From: http://trac.research.cc.gatech.edu/fastlab/ticket/302

**Comment by shrit** [comment]
Probably we will close this one before mlpack 5.0

**Comment by rcurtin** [comment]
Agreed, I think all the documentation-related refactoring and other work going on right now will end up with all mlpack algorithms accepting a matrix type soon. :+1:

**Comment by shrit** [comment]
I think we can close this one, since we have this now on our list and we have already started working on it

**Comment by DKGH** [comment]
In which release is this expected to be available?

**Comment by rcurtin** [comment]
@DKGH it's supported by many but not all of the machine learning techniques that mlpack provides.  For instance, the documentation for the naive Bayes classifier has a [section showing how to use different element types](https://www.mlpack.org/doc/user/methods/naive_bayes_classifier.html#advanced-functionality-different-element-types).  Most of the other documented techniques have similar sections in the 'advanced functionality' parts near the bottom.

**Comment by DKGH** [comment]
> [@DKGH](https://github.com/DKGH) it's supported by many but not all of the machine learning techniques that mlpack provides. For instance, the documentation for the naive Bayes classifier has a [section showing how to use different element types](https://www.mlpack.org/doc/user/methods/naive_bayes_classifier.html#advanced-functionality-different-element-types). Most of the other documented techniques have similar sections in the 'advanced functionality' parts near the bottom.

I was looking to do a standard normalization. Didn't work with arma::fmat. Worked with arma::mat.
